src/vm/threadsuspend.cpp

```diff
@@ -8231,6 +8247,126 @@ void ThreadSuspend::SuspendEE(SUSPEND_REASON reason)
 #endif //TIME_SUSPEND
 }
 
+#if defined(FEATURE_HIJACK) && defined(PLATFORM_UNIX)
+
+// This function is called when a GC is pending. It tries to ensure that the current
+// thread is taken to a GC-safe place as quickly as possible. It does this by doing 
+// one of the following:
+//
+//     - If the thread is in native code or preemptive GC is not disabled, there's
+//       nothing to do, so we return.
+//
+//     - If the thread is in interruptible managed code, we will push a frame that
+//       has information about the context that was interrupted and then switch to
+//       premptive GC mode so that the pending GC can proceed, and then switch back.
+//
+//     - If the thread is in uninterruptible managed code, we will patch the return
+//       address to take the thread to the appropriate stub (based on the return 
+//       type of the method) which will then handle preparing the thread for GC.
+//
+void PALAPI HandleGCSuspensionForInterruptedThread(CONTEXT *interruptedContext)
+{
+    Thread *pThread = GetThread();
+
+    if (pThread->PreemptiveGCDisabled() != TRUE)
+        return;
+
+    PCODE ip = GetIP(interruptedContext);
+
+    if (ExecutionManager::IsManagedCode(ip) != TRUE)
+        return;
+
+    Thread::WorkingOnThreadContextHolder workingOnThreadContext(pThread);
+    if (!workingOnThreadContext.Acquired())
+        return;
+
+    EECodeInfo codeInfo(ip);
+    if (!codeInfo.IsValid())
+        return;
+
+    DWORD addrOffset = codeInfo.GetRelOffset();
+
+    ICodeManager *pEECM = codeInfo.GetCodeManager();
+    _ASSERTE(pEECM != NULL);
+
+    bool isAtSafePoint = pEECM->IsGcSafe(&codeInfo, addrOffset);
+    if (isAtSafePoint)
+    {
+        // If the thread is at a GC safe point, push a RedirectedThreadFrame with
+        // the interrupted context and pulse the GC mode so that GC can proceed.
+        FrameWithCookie<RedirectedThreadFrame> frame(interruptedContext);
+        pThread->SetSavedRedirectContext(NULL);
+
+        frame.Push(pThread);
+
+        pThread->PulseGCMode();
+
+        frame.Pop(pThread);
+    }
+    else
+    {
+        // The thread is in non-interruptible code.
+        ExecutionState executionState;
+        StackWalkAction action;
+        REGDISPLAY regDisplay;
+        pThread->InitRegDisplay(&regDisplay, interruptedContext, true /* validContext */);
+
+        if (!pThread->IsSafeToInjectThreadAbort(interruptedContext))
```

Comment:
I've replaced this with a call to IsIPInEpilog (which was the real intention behind calling this here anyway).